File: 240415/거울에 레이저 쏘기 2/shoot-a-laser-in-the-mirror-2.py
N = int(input())
board = [list(input()) for _ in range(N)]
K = int(input())

dx, dy = [1, 0, -1, 0], [0, -1, 0, 1]

slash_dir = {
    0: 1,
    1: 0,
    2: 3,
    3: 2
}

rev_slash_dir = {
    0: 3,
    1: 2,
    2: 1,
    3: 0
}

lazer_start_dir = 0
r, c = 0, 0

if K <= N:
    lazer_start_dir = 0
    r, c = 0, K - 1
elif N < K and K <= 2 * N:
    lazer_start_dir = 1
    r, c = K - N - 1, N - 1
elif 2 * N < K and K <= 3 * N:
    lazer_start_dir = 2
    r, c = N - 1, 3 * N - K
elif K > 3 * N:
    lazer_start_dir = 3
    r, c = 4 * N - K, 0

cnts = 0
while 0 <= r < N and 0 <= c < N:
    if board[r][c] == '/':
        lazer_start_dir = slash_dir[lazer_start_dir]
    elif board[r][c] == '\\':
        lazer_start_dir = rev_slash_dir[lazer_start_dir]

    r, c = r + dx[lazer_start_dir], c + dy[lazer_start_dir]
    cnts += 1

print(cnts)

File: 260208/거울에 레이저 쏘기 2/shoot-a-laser-in-the-mirror-2.cpp
#include <bits/stdc++.h>
using namespace std;


int main() 
{
    int N;
    cin >> N;

    vector<string> board(N);
    for (int i = 0; i < N; i++) 
    {
        cin >> board[i];
    }

    long long K;
    cin >> K;

    // Please write your code here.
    int dx[4] = {1, 0, -1, 0};
    int dy[4] = {0, -1, 0, 1};

    int slash_dir[4] = {1, 0, 3, 2};
    int rev_slash_dir[4] = {3, 2, 1, 0};

    int dir = 0;
    long long r = 0, c = 0;

    if (K <= N) 
    {
        dir = 0;
        r = 0; c = K - 1;
    } 
    else if (N < K && K <= 2LL * N) 
    {
        dir = 1;
        r = K - N - 1; c = N - 1;
    } 
    else if (2LL * N < K && K <= 3LL * N) 
    {
        dir = 2;
        r = N - 1; c = 3LL * N - K;
    } 
    else 
    { 
        dir = 3;
        r = 4LL * N - K; c = 0;
    }

    long long cnts = 0;
    while (0 <= r && r < N && 0 <= c && c < N) 
    {
        if (board[(int)r][(int)c] == '/') 
        {
            dir = slash_dir[dir];
        } 
        else if (board[(int)r][(int)c] == '\\') 
        {
            dir = rev_slash_dir[dir];
        }

        r += dx[dir];
        c += dy[dir];
        cnts++;
    }

    cout << cnts << "\n";
    return 0;
}
